reprounzip-singularity/Singularity_unpacker.py

```diff
@@ -0,0 +1,236 @@
+    cmd += "SCIF_APPOUTPUT=\"/scif/data/{}/output\"\n".format(app_name)
+    cmd += "export SCIF_APPDATA SCIF_APPNAME SCIF_APPROOT SCIF_APPMETA SCIF_APPINPUT SCIF_APPOUTPUT SCIF_DATA\n"
+    with open(os.path.join(APP_ENV_DIR, "01-base.sh"), 'w') as f:
+        f.write(cmd)
+
+# Check if bin is present in the tar if not add bin and sh
+def copy_busybox(tar):
+    if not "bin" in tar.getnames():
+        print("bin is absent")
+        tar.add("../../bin",arcname="bin")
+    elif not "sh" in tar.getnames():
+        print("bin but no sh")
+        tar.add("../../bin/sh", arcname='bin/sh')
+
+
+def deleteContent(fName):
+    with open(fName, "w"):
+        pass
+
+def add_singularity_folder(tar):
+    # check if the runs are multiple or single
+    my_dict = yaml.load(open(source_config))
+    runs = my_dict['runs']
+    copy_busybox(tar)
+    deleteContent(os.path.join(ENV_DIR, "94-appsbase.sh"))
+    if len(runs) > 1:
+	print("multiple runs")
+        # App structure is present
+        # First add the SCIF folder 
+        tar.add(SCIF_DIR, arcname="scif")
+        for run in runs:
+            binary = run['binary']
+            workingdir = run['workingdir']
+            run_file = run['argv'][1]
+            app_name = run['id']
+            make_runscript(workingdir, binary, run_file, apps=True)
+            write_env_file(runs[0].get('environ'), RUN_ENV_FILE, apps=True)
+            make_app_specific_base_script(app_name)
+            make_main_app_base_script(app_name)
+            # add the modified app.
+            tar.add(APP_DIR, arcname="scif/apps/" + app_name)
+
+    else:
+        write_env_file(runs[0].get('environ'),RUN_ENV_FILE)
+        make_runscript(runs[0]['workingdir'], runs[0]['binary'], runs[0]['argv'][1])
+    tar.add(SINGULARITY_DIR, arcname=".singularity.d")
+    
+
+
+def setup(filename):
+    rpz = tarfile.open(filename, 'r:*')
+
+    # Open the inner tar in the original, without extracting it to disk
+    data = rpz.extractfile('DATA.tar.gz')
+    tar = tarfile.open('DATA.tar.gz', fileobj=data)
+
+    # Open the new tar we're writing
+    new = tarfile.open('new.tar.gz', 'w:gz')
+    # For each member of the data tar
+    for info in tar.getmembers():
+        # Make a new TarInfo, removing the DATA/ prefix from the file name
+        new_info = copy.copy(info)
+        new_info.name = info.name[5:]
+        if new_info.name:
+            # Copy the file from the inner tar to the new tar
+            if new_info.isreg():
+                new.addfile(new_info, tar.extractfile(info.name))
+            elif new_info.isdir():
+                new.addfile(new_info)
+            else:
+                new.addfile(new_info)
+    # Add the missing folders - proc,run, sys and temp_home
+    folders = ['proc','dev','sys','temp_home','mnt']
+    for folder in folders:
+        new.add("../../missing_folders/"+folder,folder)
+    print("added missing folder")
+
+    rpz.extractall()
```

Comment:
The whole point of the code above that I sent you is to not extract the files to disk. Remember that file permissions will not round-trip (also it uses at least twice as much disk space).